### ISSUE TITLE
Readonly arrays in interfaces don't allow dynamically adding data

### ISSUE THREAD

**Issue by ImRodry** [discordjs/discord.js#10339]
### Which package is this bug report for?

discord.js

### Issue description

The change in #10045 made it so that if you have a variable typed as one of the interfaces affected by the PR (will use BaseMessageOptions as an example) you will not be able to perform operations like push to your components, embeds, files and stickers arrays, even if you declare those arrays yourself and know they're not readonly. The goal of this change was to allow both readonly and mutable arrays to be passed, which does happen but only if you pass an array directly to the property, after that it is treated as readonly regardless. This was also discussed in https://discord.com/channels/222078108977594368/1249001503742758984
I don't think we're gaining anything from this change and would like to propose reverting the change on interfaces (while keeping it on methods)

### Code sample

```typescript
const res: InteractionReplyOptions = {}
res.embeds = []
res.embeds.push(new EmbedBuilder({description: "Test" })) //This errors because you cannot .push on a readonly array
// This is a simplified code sample
```

### Versions

14.15.3

### Issue priority

Medium (should be fixed soon)

### Which partials do you have configured?

Not applicable

### Which gateway intents are you subscribing to?

Not applicable

**Comment by ImRodry** [comment]
@Qjuh can we get an explanation as to why this isn't planned?

**Comment by Qjuh** [comment]
The interfaces in question are made to be the types of the parameters in methods. So keeping readonly on the methods but not the interfaces is not going to happen. If you *need* to make a type for your code that doesn't have readonly without rewriting the whole type of discord.js I propose to use a helper type like 
```ts
type NoReadonly<Type> = {
    [Key in keyof Type]: Type[Key] extends readonly (infer Content)[] ? Content[] : Type[Key];
}
```
and type your variable as `NoReadonly<InteractionReplyOptions>` or give it a shorter alias name in your typings.

**Comment by didinele** [comment]
or
```ts
type NoReadonly<Type> = { -readonly [Key in keyof Type]: Type[Key] };
```

**Comment by ImRodry** [comment]
That feels like a very unnecessary workaround, the types you export should be usable out of the box without such modifications

**Comment by vladfrangu** [comment]
The main reason our types are laid out like they are is because otherwise you'd get errors if you pass in a readonly array in places that request arrays. Now, that does cause issues if you want to do something similar to what you wanna achieve, and maybe we can eventually consider types that can help, but as of right now it just is this way

**Comment by didinele** [comment]
No, I don't think so. This type is defined since it's a complex input we receive in a function, it is:
1) helpful for us to not inline it into the definition for readability
2) allows reuse both internally and for end users

I think that as a library, it is not our job to provide the same type but with a weaker typing just because it is convenient for some random use case.

You have a multitude of ways to go about this, none of them are "workarounds", just necessary work for what you're trying to do.
